### PR TITLE
Production pull adding quality control endpoints and fixing an old CI/CD issue.

### PR DESCRIPTION
**Pull request by bubby932**
* Added middleware for authenticating a token with an arbitrary tag.
* Updated eslintrc.json to be stricter about async/await.
* Added Test Case Management endpoints:
     * /dev/quality-control/test-cases
     * /dev/quality-control/test-case/:_id/relinquish
     * /dev/quality-control/test-case/:_id/assign-self
     * /dev/quality-control/test-case/:_id/set-status/:active
     * /dev/quality-control/submit-test-case
* Bumped `firebase` from ^9.11.0 to ^9.12.1
* Bumped `express` from ^4.17.1 to ^4.18.2
* Fixed an issue that would throw an error when encountering an undefined authentication header on /dev/pull-origin.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
         "no-invalid-this": "error",
         "semi-style": "error",
         "semi-spacing": "error",
-        "no-useless-escape": "off"
+        "no-useless-escape": "off",
+        "no-await-in-loop": "error"
```
Comment (bubby932) [inline .eslintrc.json:20]: Removes some ambiguity in loops.

File: middleware.js [hunk 2]
```diff
@@ -5,7 +5,8 @@ module.exports = {
     authenticateToken: authenticateToken,
     authenticateDeveloperToken: authenticateDeveloperToken,
     authenticateToken_internal: authenticateToken_internal,
-    authenticateToken_optional: authenticateToken_optional
+    authenticateToken_optional: authenticateToken_optional,
+    authenticateTokenAndTag: authenticateTokenAndTag
```
Comment (bubby932) [inline middleware.js:9]: Helpful middleware for authenticating a user with an arbitrary tag.

File: middleware.js [hunk 3]
```diff
@@ -81,6 +82,41 @@ async function authenticateDeveloperToken(req, res, next) {
     }
 }
 
+function authenticateTokenAndTag(tag) {
```
Comment (bubby932) [inline middleware.js:85]: See my previous comment.

File: package.json [hunk 4]
```diff
@@ -16,11 +16,11 @@
   "dependencies": {
     "bcryptjs": "^2.4.3",
     "dotenv": "^16.0.3",
-    "express": "^4.17.1",
+    "express": "^4.18.2",
```
Comment (bubby932) [inline package.json:19]: Suggested by @dependabot

File: package.json [hunk 5]
```diff
@@ -16,11 +16,11 @@
...
     "express-fileupload": "^1.2.1",
     "express-rate-limit": "^6.6.0",
     "express-slow-down": "^1.5.0",
-    "firebase": "^9.11.0",
+    "firebase": "^9.12.1",
```
Comment (bubby932) [inline package.json:23]: Suggested by @dependabot

File: routers/dev.js [hunk 6]
```diff
@@ -3,6 +3,8 @@ const { PullPlayerData, PushPlayerData } = require('../helpers');
 const middleware = require('../middleware');
 const config = require('../config.json');
 const { execSync } = require('node:child_process');
+const { authenticateTokenAndTag } = require('../middleware');
```
Comment (bubby932) [inline routers/dev.js:6]: New dependencies for QC endpoints.

File: routers/dev.js [hunk 7]
```diff
@@ -56,7 +58,7 @@ router.post("/pull-origin", async (req, res) => {
         let Authentication = req.headers.authorization.split(' ')[1];
         let key = config.development_mode ? process.env.DEV_PULL_SECRET : process.env.PRODUCTION_PULL_SECRET;
 
-        if(Authentication.length != key.length || Authentication != key) return res.status(403).json({
+        if(Authentication?.length != key?.length || Authentication != key) return res.status(403).json({
```
Comment (bubby932) [inline routers/dev.js:61]: Fix CI issue.

File: routers/dev.js [hunk 8]
```diff
@@ -80,6 +82,216 @@ router.post("/pull-origin", async (req, res) => {
     }
 });
 
+router.get("/quality-control/test-cases", authenticateTokenAndTag("QA Tester"), async (req, res) => {
```
Comment (bubby932) [inline routers/dev.js:85]: Fetch test case endpoint, with filters.

File: routers/dev.js [hunk 9]
```diff
@@ -80,6 +82,216 @@ router.post("/pull-origin", async (req, res) => {
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred and we were unable to serve your request. Please inform the API team."
+        });
+        throw ex;
+    }
+});
+
+router.post("/quality-control/test-case/:_id/relinquish", authenticateTokenAndTag("QA Tester"), async (req, res) => {
```
Comment (bubby932) [inline routers/dev.js:118]: Relinquishing assignment of a test case.

File: routers/dev.js [hunk 10]
```diff
@@ -80,6 +82,216 @@ router.post("/pull-origin", async (req, res) => {
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred. Please notify the API team immediately."
+        });
+        throw ex;
+    }
+});
+
+router.post("/quality-control/test-case/:_id/assign-self", authenticateTokenAndTag("QA Tester"), async (req, res) => {
```
Comment (bubby932) [inline routers/dev.js:166]: Taking assignment of a test case.

File: routers/dev.js [hunk 11]
```diff
@@ -80,6 +82,216 @@ router.post("/pull-origin", async (req, res) => {
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred and we were unable to serve your request. Please alert the API team."
+        });
+        throw ex;
+    }
+});
+
+router.post("/quality-control/test-case/:_id/set-status/:active", authenticateTokenAndTag("QA Tester"), async (req, res) => {
```
Comment (bubby932) [inline routers/dev.js:214]: Setting the status of a test case.

File: routers/dev.js [hunk 12]
```diff
@@ -80,6 +82,216 @@ router.post("/pull-origin", async (req, res) => {
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal server error occurred and we could not serve your request. Please notify the API team."
+        });
+        throw ex;
+    }
+});
+
+router.put("/quality-control/submit-test-case", authenticateTokenAndTag("QA Tester"), async (req, res) => {
```
Comment (bubby932) [inline routers/dev.js:257]: Creating a new test case.

File: routers/matchmaking.js [hunk 13]
```diff
@@ -84,7 +84,7 @@ class RoomSession {
     MaxPlayers; // The maximum number of players the instance can contain.
     Age; // Age of the instance.
     AgeWithoutPlayer;
-    TTL = 1;
+    TTL = 300*1000;
```
Comment (bubby932) [inline routers/matchmaking.js:87]: Fixed a TTL default issue.

File: routers/ws/WebSocketServerV2.js [hunk 14]
```diff
@@ -677,9 +677,9 @@ WebSocketServerV2.on('connection', (Socket) => {
         delete ws_connected_clients[ConnectedUserData.uid];
         auditLog(`User "${ConnectedUserData.nickname}" / @${ConnectedUserData.username} with ID ${ConnectedUserData.uid} has disconnected. Currently online players: ${Object.keys(ws_connected_clients).length}`);
 
-        if (ConnectedUserData.matchmaking_InstanceId !== null) {
+        if (ConnectedUserData.matchmaking_InstanceId != null) {
```
Comment (bubby932) [inline routers/ws/WebSocketServerV2.js:680]: Fixed some JS ambiguity.

**Review by bubby932 — COMMENTED** [review]
Explained every change in more detail.

**Comment by bubby932** [comment]
<@363450765307805696> review please 🥺

**Review by Bobrobot1 — APPROVED** [review]
looks good, 300 * 1000 kinda unnecessary operation tho 😬

**Pull request merged**